contracts/utils/Swap.sol

```diff
@@ -0,0 +1,454 @@
+        require(offer.isOpen, "Offer already closed");
+
+        for (uint256 i = 0; i < bids.length; i++) {
+            _swap(offer, bids[i]);
+        }
+
+        if (offer.availableSize == 0) {
+            offer.isOpen = false;
+
+            emit SettleOffer(swapId);
+        }
+    }
+
+    /**
+     * @notice Close offer
+     * @param swapId swapId unique identifier of the swap offer
+     */
+    function closeOffer(uint256 swapId) external override {
+        Offer storage offer = swapOffers[swapId];
+        require(offer.seller != address(0), "Offer does not exist");
+        require(msg.sender == offer.seller, "Only seller can close offer");
+        require(offer.isOpen, "Offer already closed");
+
+        offer.isOpen = false;
+
+        emit CloseOffer(swapId);
+    }
+
+    /**
+     * @notice Cancel one or more nonces
+     * @dev Cancelled nonces are marked as used
+     * @dev Emits a Cancel event
+     * @dev Out of gas may occur in arrays of length > 400
+     * @param nonces uint256[] List of nonces to cancel
+     */
+    function cancelNonce(uint256[] calldata nonces) external override {
+        for (uint256 i = 0; i < nonces.length; i++) {
+            uint256 nonce = nonces[i];
+            if (_markNonceAsUsed(msg.sender, nonce)) {
+                emit Cancel(nonce, msg.sender);
+            }
+        }
+    }
+
+    /************************************************
+     *  PUBLIC VIEW FUNCTIONS
+     ***********************************************/
+
+    /**
+     * @notice Validates Swap bid for any potential errors
+     * @param bid Bid struct containing bid details
+     * @return tuple of error count and bytes32[] memory array of error messages
+     */
+    function check(Bid calldata bid)
+        public
+        view
+        returns (uint256, bytes32[] memory)
+    {
+        Offer memory offer = swapOffers[bid.swapId];
+        require(offer.seller != address(0), "Offer does not exist");
+
+        bytes32[] memory errors = new bytes32[](MAX_ERROR_COUNT);
+
+        uint256 errCount;
+
+        // Check signature
+        address signatory = _getSignatory(bid);
+
+        if (signatory == address(0)) {
+            errors[errCount] = "SIGNATURE_INVALID";
+            errCount++;
+        }
+
+        if (signatory != bid.signerWallet) {
+            errors[errCount] = "SIGNATURE_MISMATCHED";
+            errCount++;
+        }
+
+        // Check nonce
+        if (nonceUsed(signatory, bid.nonce)) {
+            errors[errCount] = "NONCE_ALREADY_USED";
+            errCount++;
+        }
+
+        // Check bid size
+        if (bid.buyAmount < offer.minBidSize) {
+            errors[errCount] = "BID_TOO_SMALL";
+            errCount++;
+        }
+
+        // Check bid price
+        uint256 bidPrice =
+            uint256(bid.sellAmount)
+                .mul(uint256(10)**offer.offeredTokenDecimals)
+                .div(bid.buyAmount);
```

Comment:
Since overflow/underflow checks are already included by default in Solidity, we can avoid using the SafeMath library entirely https://soliditydeveloper.com/solidity-0.8